### PR TITLE
✨Add Letra-MNC-FY21 constants to kflib

### PR DESCRIPTION
**Pull request by r3m0chop**
New ingest constants as for the Letra-MNC-FY21 clinical data, thank you very much!
cc: M1 team: @znatty22 , @chris-s-friedman , @youngnm , Meen, etc
cc: KF-DRC: @Christina-J-Diaz , @HigginsD3b, @aw3334

### CODE REVIEW

File: kf_lib_data_ingest/common/constants.py
Comment (chris-s-friedman) [inline kf_lib_data_ingest/common/constants.py:407]: I think it would make sense to specify if the half-siblings are paternal or maternal. Please add Paternal and Maternal entries for each

**Review by chris-s-friedman — CHANGES_REQUESTED** [review]
I think the procurement and anatomy site values make sense. 

For family relationships, please add: 

1. the appropriate reverse relationships (i.e. [here](https://github.com/kids-first/kf-lib-data-ingest/blob/123ae7b686b71f56abfe3907fe5d5fcff71e1ba4/kf_lib_data_ingest/common/constants.py#L429-L463)).
2. the appropriate gender-relationship mapping ([here](https://github.com/kids-first/kf-lib-data-ingest/blob/123ae7b686b71f56abfe3907fe5d5fcff71e1ba4/kf_lib_data_ingest/common/constants.py#L473-L504))
3. the appropriate gendered-relationship mapping ([here](https://github.com/kids-first/kf-lib-data-ingest/blob/123ae7b686b71f56abfe3907fe5d5fcff71e1ba4/kf_lib_data_ingest/common/constants.py#L507-L552))
4. the appropriate sibling-relationship mappings ([here](https://github.com/kids-first/kf-lib-data-ingest/blob/123ae7b686b71f56abfe3907fe5d5fcff71e1ba4/kf_lib_data_ingest/common/constants.py#L556-L582))

File: kf_lib_data_ingest/common/constants.py
Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py:407]: Fully agreed, thanks for the catch!  

In addition to the 4x other areas to add any such new relationships, should I also add them to:
5. the appropriate relationship-parent mappings ([here](https://github.com/kids-first/kf-lib-data-ingest/blob/123ae7b686b71f56abfe3907fe5d5fcff71e1ba4/kf_lib_data_ingest/common/constants.py#L556-L581))
???

Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py:407]: Scratch that about #5, since I now better see that is the same as #4.

Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py:407]: @chris-s-friedman , et al: for a genderless reverse relationship from Aunts and Uncles..... shall we use "Niblings"?
as per: https://www.merriam-webster.com/words-at-play/words-were-watching-nibling

Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py:407]: In discussing further with the KF-DRC team, wondering if "nibling" and "pibling" could be used as the default genderless versions (of niece/nephew, and aunt/uncle, specifically), with the gendered reversals defined to help refine these to the gender specific version.

Comment (Christina-J-Diaz) [reply kf_lib_data_ingest/common/constants.py:407]: I am wondering if the reverse relationships _need_ to be genderless for nephew/niece and aunt/uncle. I believe when we say reverse relationship we're referring to populating the value of `participant2_to_participant1_relation` ? it would make sense for this to be left as niece/nephew or aunt/uncle... ?

Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py:407]: Well, at the first pass, prior to the kf-ingest-data-ingest bringing in the gender and refining it where possible, (much like Sibling is set as the opposite of both Brother and Sister, and Child as the opposite of both Father and Mother) this would allow the first-pass default of a genderless value, should gender not exist on the participant for whatever reason.

Comment (Christina-J-Diaz) [reply kf_lib_data_ingest/common/constants.py:407]: Yea I guess I'm wondering in what circumstances would we have a participant without a sex and yet have an indicated relationship? And specifically wondering how that situation would arise with aunt/uncle and niece/nephew? If the submitted relationship is "niece", "nephew", "aunt", "uncle" the given sex of subject should align, and if it doesn't isnt this something we would bring up to the X01 group and have corrected?

Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py:407]: Thanks for the thoughts, @Christina-J-Diaz .  

Even if these particular participants in this clinical data set should all have their sex, it seems that this kf-lib-data-ingest system is generally setup for the genderless default to be the first pass value before it then performs any such gender refinement.  As such, it does feel to me that to not include something there (even if it's a new one to us, such as "nibling" and "pibling") as its first pass default might be remiss.  Doesn't adding these genderless reverse lookups align with the spirit of the kf-lib-data-ingest system?  Thanks!

Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py:407]: Another thought on this.... if the terms "nibling" and "pibling" in particular are of concern, what about just:
* "Niece or Nephew"
* "Aunt or Uncle"

or perhaps even a hybrid combination, such as:
* "Nibling (Niece or Nephew)"
* "Pibling (Aunt or Uncle)"
?

Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py]: For your consideration, I have committed potential refinements (https://github.com/kids-first/kf-lib-data-ingest/pull/657/commits/a09004acecf995df1c83b19904f8a989de26fb05) to demonstrate the possibilities of such hybrid combinations, e.g.,:
> PIBLING = "Pibling (Aunt or Uncle)"
> NIBLING = "Nibling (Niece or Nephew)"

**Comment by r3m0chop** [comment]
commits squashed :-)

File: kf_lib_data_ingest/common/constants.py
Comment (Christina-J-Diaz) [reply kf_lib_data_ingest/common/constants.py]: @r3m0chop Maybe we can use "other" for right now? Since this will be default behavior, it will affect all ingest packages--including other teams that may use the packages (like the M1 team). Maybe we can default to "other" temporarily and then have a larger discussion about this in a separate PR?

Comment (r3m0chop) [reply kf_lib_data_ingest/common/constants.py]: Thank you, @Christina-J-Diaz .  

Howzabout the following for such initial "other" text description placeholders (noting the internal constant name shouldn't be visible outside of our code):

*     MATERNAL_PIBLING = "Maternal Other (Aunt or Uncle)"
*     PATERNAL_PIBLING = "Paternal Other (Aunt or Uncle)"
*     PIBLING = "Other (Aunt or Uncle)"
*     MATERNAL_NIBLING = "Maternal Other (Niece or Nephew)"
*     PATERNAL_NIBLING = "Paternal Other (Niece or Nephew)"
 *    NIBLING = "Other (Niece or Nephew)"
etc

Comment (Christina-J-Diaz) [reply kf_lib_data_ingest/common/constants.py]: @r3m0chop I like this. I think it is a good compromise for a placeholder. I appreciate the additional "maternal" / "paternal" description with other-- that is more informative than leaving as "other". I'm not 100% sold on the parentheses (because its a much longer than string compared to the other values, but I don't see an issue with it for now 😄 )

Comment (Christina-J-Diaz) [reply kf_lib_data_ingest/common/constants.py]: This might have an obvious answer but thinking about vocabulary in terms of pedigrees -- do we need the maternal / paternal distinction for niece / nephew? Biologically you only have one set of nieces/nephews , which are all the children of your sibling

Comment (chris-s-friedman) [reply kf_lib_data_ingest/common/constants.py]: Great thoughts, folks. just to build on this, saying `Other (some stuff in parens)` is confusing - what about some other *other* relation besides aunt/uncle, niece/ nephew? 

How about: 

> MATERNAL_PIBLING = "Maternal Aunt or Uncle"
PATERNAL_PIBLING = "Paternal Aunt or Uncle"
PIBLING = "Aunt or Uncle"
MATERNAL_NIBLING = "Maternal Niece or Nephew"
PATERNAL_NIBLING = "Paternal Niece or Nephew"
NIBLING = "Niece or Nephew"

Comment (Christina-J-Diaz) [reply kf_lib_data_ingest/common/constants.py]: Great point. Agreed -- this looks cleaner. @r3m0chop what do you think? I am actually appreciating that the constant name is PIBLING or NIBLING and the value is the _definition_ of pibling/nibling.

**Comment by r3m0chop** [comment]
commit of latest pibling/nibling updates have been squashed, and this might therefore be good to go, yeah?
cc: @chris-s-friedman & @Christina-J-Diaz 
Thank you both for looking at this, and all the great suggestions and dialogue. 👍

**Comment by r3m0chop** [comment]
Sorry, but I've just found a last minute update.... in beginning to prepare the Letra ingest to use these new constants, I just encountered a previously unseen new value for Anatomy Site that must have been quietly introduced in a latest clinical data revision.  

I first want to discuss with the rest of the DRC team, as the additional new constant would be:
* `SPECIMEN.ANATOMY_SITE.UMBILICAL_CORD == "Umbilical Cord"`

**Review by Christina-J-Diaz — COMMENTED** [review]
Looks good! Just need to add the last constant for anatomy site 👍

**Review by Christina-J-Diaz — APPROVED** [review]
Great!

**Comment by r3m0chop** [comment]
Awesome, @Christina-J-Diaz .  I have added `SPECIMEN.ANATOMY_SITE.UMBILICAL_CORD` in a new commit, but will await approval from @HigginsD3b before squashing the commits.  Thanks!

**Comment by r3m0chop** [comment]
@HigginsD3b says:
> _I think it makes sense._
so I will do one last squash on these commits momentarily....

**Comment by r3m0chop** [comment]
@chris-s-friedman - how's this all looking to you now?  Thanks!

**Review by Christina-J-Diaz — APPROVED** [review]

**Review by znatty22 — COMMENTED** [review]
This is a general comment for this PR. Wondering if these constants appear in other studies/datasets or just Letra? I think the intended purpose of the constants file was to cover the common terms not necessarily all of the terms, so just wondering how do we decide to add something as a constant to the ingest lib versus just use the string literal in the ingest package's extract config (`constants.ARM` versus "arm")

**Comment by r3m0chop** [comment]
Following a quick stand-up chat this AM, it sounds like the next steps here might be:
1. ensure that @chris-s-friedman is okay with the final revised+squashed commit, and await his Approval on this PR
2. once approved, I can perform the Merge PR
3. and then ping @znatty22 and @Christina-J-Diaz to cut a release
yeah?

**Comment by r3m0chop** [comment]
Hey, @chris-s-friedman, I'm reviving this PR and would like to ask if you are okay with the final revised+squashed commit, to Approve this PR?

We recently learned that `CONCEPT.FAMILY_RELATIONSHIP.RELATION_FROM_1_TO_2` is not as flexible as other less complex concepts in their use of values which are not already defined here as constants.

For example, given the following sample `value_map()`:

```
value_map(
    in_col="FamilyRelationship",
    m={
        # TYPE A: Family Relationships with pre-defined constants:
        #
        "Paternal Grandfather": constants.RELATIONSHIP.PATERNAL_GRANDFATHER,

        ################################################
​
        # TYPE B: Family Relationships without pre-defined constants,
        # instead using hard-coded pass-through values
        #
        # TODO: Once each of hard-coded descriptor has been PR merge'd into
		# kf_lib_data_ingest/common/constants.py,
        # go ahead and convert the following into the newly corresponding constants.
        # Until such a PR is approved and merged into kf_lib_data_ingest, though,
        # list each outstanding relationship type below with hardcoded pass-through strings:
        #
        "Paternal Great-grandfather": "Paternal Great-grandfather",
    },
    #
    # NOTE: this study will not be using this as RELATIONSHIP_TO_PROBAND, but as a more general 
    # FAMILY_RELATIONSHIP.RELATION_FROM_1_TO_2 between any 2x participants,
    # 1x of which might NOT be a Proband:
    #
    out_col=CONCEPT.FAMILY_RELATIONSHIP.RELATION_FROM_1_TO_2,
​
    # [...]
)
```

* TYPE A: (constants) 
...generate well balanced pairs of relationships, such as:

| participant1_id | participant2_id | participant1_to_participant2_relation |
| ---- | ------- | ----------- |
| PERSON_A | PERSON_B | Paternal Grandchild |
| PERSON_B | PERSON_A | Paternal Grandfather |

* TYPE B:  (pass-through strings)
...utilize the same value for both sides of the relationship, such as:

| participant1_id | participant2_id | participant1_to_participant2_relation |
| ---- | ------- | ----------- |
| PERSON_C | PERSON_D | Maternal Uncle |
| PERSON_D | PERSON_C | Maternal Uncle |


This does make sense, as the ingest library has no pre-conceived notion of any pass-through strings.

As per:
https://github.com/kids-first/kf-lib-data-ingest/blob/master/kf_lib_data_ingest/common/family_relationships.py#L172-L173

```
    reverse_df[FR.RELATION_FROM_1_TO_2].replace(
        constants.REVERSE_RELATIONSHIPS, inplace=True
```

.. the `_bidirect_relationships()` method uses a `.replace()` ([rather than a map()](https://stackoverflow.com/questions/43367700/python-dataframe-replace-values-using-dictionary-convert-nan-if-not-in-diction)) which aligns with the observed results.

As such, it looks like we'll need to get these new constants into `kf-lib-data-ingest` after all.  Thank you! :-)

**Review by chris-s-friedman — APPROVED** [review]
I think all of this makes sense to me. I appreciate you putting all the time and effort into this!

Like I think you noted, I think we should get @znatty22 to approve on this before merging this PR.

**Review by znatty22 — APPROVED** [review]
✅ Thanks for taking a good look into this @r3m0chop! Approving but also want to leave this note here in case you want to try it before mering:

You can always test out your changes to the ingest lib (in an ingest package for example) before it gets released by updating your requirements file to look like this:
```
kf_lib_data_ingest @ git+https://github.com/kids-first/kf-lib-data-ingest@add-letra-constants-to-kflib
``` 
And then of course you'll need to `pip install -r requirements` in your venv or wherever you're using ingest lib

**Pull request merged**
